enhancements/sdk-external-and-pluggable-validations.md

```diff
@@ -0,0 +1,330 @@
+
+The validation executable will need to accept a single input value which
+is the bundle directory. The output will need to be a ManifestResult JSON. This
+JSON will be parsed by the operator-sdk converted and output as a Result.
+
+For example, the existing validations could easily be wrapped with a `main.go`
+file and compiled into a binary. They could be copied to their own repos for
+easier release. Now you might be thinking that this means we'll have many
+releases still, we actually would only have one (1) release which is the
+validator itself. For go based validators this might mean a new binary, but you
+could also release your validator in python and make it a single file or even a
+shell script.
+
+From the operator-sdk's point of view, we don't care what you use
+to create your validator only that we can run it and pass it a bundle directory.
+
+#### Validators
+
+##### Running validators
+
+The [existing validators](#default-validators-run-by-operator-sdk) operate
+on a bundle. Each validator should be some executable that accepts a
+single argument, the bundle root directory. So `operator-sdk` will pass them
+the bundle directory. It will be the validator's responsibility to parse
+the bundle to get what it needs.
+
+For example, the validator executable will be invoked from the
+`operator-sdk` in the following manner:
+
+```
+/path/to/validator/executable /path/to/bundle
+```
+
+A concrete example might look like this:
+
+```
+./validator-poc /home/user/dev/gatekeeper-operator/bundle
+```
+
+The actual implementation is up to the author. Here we have an example of
+the [`OperatorHubValidator`][validator-poc1] as a Go binary.
+
+##### Validator results
+
+As stated earlier, each validator should be some executable that accepts a
+single argument, the bundle root directory. Th validators should also output
+ManifestResult JSON to stdout.
+
+Because the [existing validators](#default-validators-run-by-operator-sdk)
+currently return a [ManifestResult][manifest-result], it seems logical that we
+use the same object as JSON for the output.
+
+The validator executable should exit non-zero ONLY if the entrypoint failed to
+run NOT if the bundle validation fails.
+
+For example, let's say the validator is given a path to the gatekeeper bundle.
+The validator should validate the given bundle and output the ManifestResult JSON.
+Here is an example of a run:
+
+```json
+{
+    "Name": "gatekeeper-operator.v0.2.0-rc.3",
+    "Errors": null,
+    "Warnings": [
+        {
+            "Type": "CSVFileNotValid",
+            "Level": "Warning",
+            "Field": "",
+            "BadValue": "",
+            "Detail": "(gatekeeper-operator.v0.2.0-rc.3) csv.Spec.minKubeVersion is not informed. It is recommended you provide this information. Otherwise, it would mean that your operator project can be distributed and installed in any cluster version available, which is not necessarily the case for all projects."
+        }
+    ]
+}
+```
```

Comment:
Are we going to expose this as a go type that validator implementers will use?

Reply:
@ryantking yes, we are going to expose it. There is a link to it in the EP: https://github.com/operator-framework/api/blob/master/pkg/validation/errors/error.go#L9-L16

That's already accessible from SDK.

Reply:
Let me know if I need to call that out better.